Projects/Authentication/Sources/AuthenticationStore.swift

```diff
@@ -0,0 +1,215 @@
+import Apollo
+import Flow
+import Foundation
+import Presentation
+import UIKit
+import hCore
+import hCoreUI
+import hGraphQL
+
+struct OTPState: StateProtocol {
+    var isLoading = false
+    var isResending = false
+    var id: String? = nil
+    var code: String = ""
+    var codeErrorMessage: String? = nil
+    var emailErrorMessage: String? = nil
+    var email: String = ""
+    var canResendAt: Date? = nil
+
+    public init() {}
+}
+
+public struct AuthenticationState: StateProtocol {
+    var otpState = OTPState()
+
+    public init() {}
+}
+
+public enum OTPStateAction: ActionProtocol {
+    case setCode(code: String)
+    case verifyCode
+    case setLoading(isLoading: Bool)
+    case setCodeError(message: String?)
+    case setEmailError(message: String?)
+    case setEmail(email: String)
+    case setID(id: String?)
+    case submitEmail
+    case reset
+    case resendCode
+    case showResentToast
+}
+
+public enum AuthenticationNavigationAction: ActionProtocol {
+    case otpCode
+    case authSuccess(accessToken: String)
+}
+
+public enum AuthenticationAction: ActionProtocol {
+    case otpStateAction(action: OTPStateAction)
+    case navigationAction(action: AuthenticationNavigationAction)
+}
+
+public final class AuthenticationStore: StateStore<AuthenticationState, AuthenticationAction> {
+    @Inject var client: ApolloClient
+
+    public override func effects(
+        _ getState: @escaping () -> AuthenticationState,
+        _ action: AuthenticationAction
+    ) -> FiniteSignal<AuthenticationAction>? {
+        if case let .otpStateAction(action: .setCode(code)) = action {
+            let generator = UIImpactFeedbackGenerator(style: .light)
+            generator.impactOccurred()
+
+            if code.count == 6 {
+                return [
+                    .otpStateAction(action: .verifyCode),
+                    .otpStateAction(action: .setLoading(isLoading: true)),
+                ]
+                .emitEachThenEnd
+            }
+        } else if case .otpStateAction(action: .verifyCode) = action {
+            let state = getState()
+
+            return
+                client.perform(
+                    mutation: GraphQL.VerifyLoginOtpAttemptMutation(
+                        id: state.otpState.id ?? "",
+                        otp: state.otpState.code
+                    )
+                )
+                .delay(by: 0.5)
+                .compactMap { data in
+                    if data.loginVerifyOtpAttempt.asVerifyOtpLoginAttemptError != nil {
+                        return .otpStateAction(
+                            action: .setCodeError(message: L10n.Login.CodeInput.ErrorMsg.codeNotValid)
+                        )
+                    } else if let success = data.loginVerifyOtpAttempt.asVerifyOtpLoginAttemptSuccess {
+                        return .navigationAction(action: .authSuccess(accessToken: success.accessToken))
```

Comment:
I wonder if we can start to disambiguate action types and have separate handlers for effects / reduction. Would that be an interesting thing to look into? I recall we thought about it earlier - do you think it would have value?